core-object.js

```diff
@@ -1,67 +1,54 @@
 'use strict';
 
-var assignProperties = require('./lib/assign-properties');
-var deprecation = require('./lib/deprecation');
+const assignProperties = require('./lib/assign-properties');
+const deprecation = require('./lib/deprecation');
 
 function needsNew() {
   throw new TypeError("Failed to construct: Please use the 'new' operator, this object constructor cannot be called as a function.");
 }
 
-function CoreObject(options) {
-  if (!(this instanceof CoreObject)) {
-    needsNew()
+class CoreObject {
+  constructor() {
+    this.init.apply(this, arguments);
   }
-  this.init(options);
-}
 
-CoreObject.prototype.init = function(options) {
-  if (options) {
-    for (var key in options) {
-      this[key] = options[key];
+  init(options) {
+    if (options) {
+      for (let key in options) {
+        this[key] = options[key];
+      }
     }
   }
-};
-
-CoreObject.extend = function(options) {
-  var constructor = this;
-
-  function Class() {
-    var length = arguments.length;
-
-    if (length === 0)      this.init();
-    else if (length === 1) this.init(arguments[0]);
-    else                   this.init.apply(this, arguments);
-  }
-
-  Class.__proto__ = CoreObject;
 
-  Class.prototype = Object.create(constructor.prototype);
+  static extend(options) {
+    class Class extends this { }
```

Comment:
I can't help but think of the notations in official documents. `{ /* This space intentionally left blank. */ }`